hyper/http20/connection.py

```diff
@@ -266,8 +266,9 @@ def request(self, method, url, body=None, headers=None):
         # If threads interleave these operations, it could result in messages
         # being sent in the wrong order, which can lead to the out-of-order
         # messages with lower stream IDs being closed prematurely.
+        stream_id = self.putrequest(method, url)
+        self.connect()
```

Comment:
Sadly, this won't work either, as noted in the comment above.

`putrequest` obtains a stream ID, but doesn't send a HEADERS frame. That means that hyper thinks the stream ID is used, but the remote peer doesn't think that yet.

This whole problem is sufficiently troublesome that I suspect it reflects a more profound problem with the way this library is structured. It is extremely not amenable to being thread-safe except at an extremely coarse level.

I wonder if we need to investigate rewriting this library with threads. =(